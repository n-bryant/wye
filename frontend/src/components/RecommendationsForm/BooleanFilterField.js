import React from "react";
import PropTypes from "prop-types";

import createClassNameHelper from "@n_bryant/classnames-helper";
import JSS_CLASS_NAME_PREFIX from "../../../lib/classNamePrefix";
import { withStyles } from "@material-ui/core/styles";

import Switch from "@material-ui/core/Switch";
import Typography from "@material-ui/core/Typography";

import styles from "./BooleanFilterField.styles";

/**
 * renders a field for electing a boolean value in the form of a Switch
 */
export const BooleanFilterField = props => {
  const classnames = BooleanFilterField.classnames(props);
  const { classes, name, type, label, setFieldValue, value } = props;
  const [checked, setChecked] = React.useState(false);

  /**
   * handle changes to the Switch,
   * and update formik field value
   * @param {String} option
   */
  const handleChange = event => {
    setChecked(event.target.checked);
    setFieldValue(name, {
      ...value,
      gameFilters: {
        ...value.gameFilters,
        [type]: event.target.checked
      }
    });
  };

  return (
    <div className={classnames.root()}>
      <Typography
        className={classnames.element("label")}
        variant="body2"
        gutterBottom={true}
      >
        {label}
      </Typography>
      <div>
        <Switch
          classes={{
            switchBase: classes.switch,
            checked: classes.switchChecked,
            track: classes.switch
          }}
          checked={checked}
          onChange={handleChange}
          value={label}
          color="primary"
          inputProps={{ "aria-label": `${type}-checkbox` }}
        />
      </div>
    </div>
  );
};
BooleanFilterField.classnames = createClassNameHelper(
  `${JSS_CLASS_NAME_PREFIX}BooleanFilterField`
);
BooleanFilterField.propTypes = {
  // styles to apply
  classes: PropTypes.shape({
    root: PropTypes.string,
    label: PropTypes.string,
    switch: PropTypes.string,
    switchChecked: PropTypes.string
  }),
  // Formik field name
  name: PropTypes.string,
  // the label to display for the field
  label: PropTypes.string,
  // the type of field
  type: PropTypes.oneOf(["freeToPlay", "onSale"]).isRequired,
  // gameFilters field value from Formik form
  value: PropTypes.object,
  // handler for updating users field value
  setFieldValue: PropTypes.func
};
BooleanFilterField.defaultProps = {
  classes: {}
};

// apply styles
export default withStyles(styles)(BooleanFilterField);
