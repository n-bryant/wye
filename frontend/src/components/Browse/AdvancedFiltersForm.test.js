import { Container } from "@material-ui/core";
import BooleanFilterField from "../RecommendationsForm/BooleanFilterField";
import AdvancedFiltersFormWithFormik, {
  getCategoryOptions,
  getMinValue,
  getMaxValue,
  AdvancedFiltersForm
} from "./AdvancedFiltersForm";

describe("AdvancedFiltersFormWithFormik", () => {
  const props = {
    initialValues: {},
    items: [
      {
        node: {
          game: {
            appid: "1",
            name: "foo",
            publishers: ["a"],
            developers: ["a"],
            genres: ["a"],
            tags: ["a"]
          }
        }
      }
    ]
  };

  const gamesFilterWidgetRenderCallbackArgs = {
    setFilterOptions: jest.fn()
  };

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(
      shallow(<AdvancedFiltersFormWithFormik {...props} />).prop("children")(
        gamesFilterWidgetRenderCallbackArgs
      )
    );
  });

  it("should render successfully", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should set some defaults for the initialValues passed to the Formik form", () => {
    expect(wrapper.prop("value").initialValues).toMatchObject({
      users: [],
      filters: {},
      orderBy: [],
      sortOrder: "DESC"
    });
  });

  it("should pass formik props and any received props to its rendered child", () => {
    expect(wrapper.prop("children").props.formik).toBeDefined();
  });
});

const games = [
  {
    appid: "1",
    discount: 25,
    finalPrice: 1000,
    ownersMin: 500,
    ownersMax: 1000,
    genres: ["a"],
    tags: ["a"]
  },
  {
    appid: "2",
    discount: 15,
    finalPrice: 500,
    ownersMin: 5000,
    ownersMax: 10000,
    genres: ["a", "b"],
    tags: ["a"]
  },
  {
    appid: "3",
    discount: 50,
    finalPrice: 350,
    ownersMin: 1000,
    ownersMax: 2000,
    genres: ["a", "b", "c"],
    tags: ["a"]
  },
  {
    appid: "4",
    discount: 75,
    finalPrice: 250,
    ownersMin: 150,
    ownersMax: 300,
    genres: ["a", "b", "c", "d"],
    tags: ["a"]
  },
  {
    appid: "5",
    discount: 90,
    finalPrice: 1500,
    ownersMin: 2500,
    ownersMax: 5000,
    genres: ["a", "b", "c", "d", "e"],
    tags: ["a"]
  }
];

describe("AdvancedFiltersForm", () => {
  const props = {
    classes: {
      root: "root",
      closing: "closing",
      drawer: "drawer",
      drawerContent: "drawerContent",
      modal: "modal",
      paper: "paper",
      backdrop: "backdrop",
      closeButtonContainer: "closeButtonContainer",
      fieldsContainer: "fieldsContainer",
      fieldSection: "fieldSection",
      fieldSectionTitle: "fieldSectionTitle",
      menuPaper: "menuPaper",
      selectOption: "selectOption",
      submitButtonContainer: "submitButtonContainer"
    },
    gameList: games,
    drawerCloseHandler: jest.fn(),
    formik: {
      setFieldValue: jest.fn(),
      values: {}
    },
    hiddenFields: []
  };

  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<AdvancedFiltersForm {...props} />);
  });

  it("should render successfully", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("should render the Player Filters section if a formik users value is defined and has length", () => {
    wrapper.setProps({
      formik: {
        ...props.formik,
        values: {
          users: ["a"]
        }
      }
    });
    wrapper.update();
    expect(
      wrapper.findWhere(n => n.text() === "Player Filters:").length
    ).not.toBe(0);
  });

  it("should hide the Free to Play and On Sale BooleanFilterFields if the hiddenFields prop includes 'onSale'", () => {
    wrapper.setProps({
      hiddenFields: ["onSale"]
    });
    wrapper.update();

    expect(wrapper.find(BooleanFilterField).length).toBe(0);
  });
});
