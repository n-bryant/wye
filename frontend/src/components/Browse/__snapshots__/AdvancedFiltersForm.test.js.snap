// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdvancedFiltersForm should render successfully 1`] = `
<ForwardRef(Drawer)
  ModalProps={
    Object {
      "BackdropProps": Object {
        "classes": Object {
          "root": "backdrop",
        },
        "invisible": true,
      },
      "disablePortal": true,
    }
  }
  anchor="right"
  className="wye-AdvancedFiltersForm root"
  classes={
    Object {
      "docked": "MuiDrawer-docked",
      "modal": "MuiDrawer-modal modal",
      "paper": "MuiDrawer-paper",
      "paperAnchorBottom": "MuiDrawer-paperAnchorBottom",
      "paperAnchorDockedBottom": "MuiDrawer-paperAnchorDockedBottom",
      "paperAnchorDockedLeft": "MuiDrawer-paperAnchorDockedLeft",
      "paperAnchorDockedRight": "MuiDrawer-paperAnchorDockedRight",
      "paperAnchorDockedTop": "MuiDrawer-paperAnchorDockedTop",
      "paperAnchorLeft": "MuiDrawer-paperAnchorLeft",
      "paperAnchorRight": "MuiDrawer-paperAnchorRight paper",
      "paperAnchorTop": "MuiDrawer-paperAnchorTop",
      "root": "MuiDrawer-root drawer",
    }
  }
  onClose={[Function]}
  open={true}
>
  <WithStyles(ForwardRef(Container))
    className="wye-AdvancedFiltersForm__drawerContent drawerContent"
    disableGutters={true}
  >
    <Styled(MuiBox)
      className="wye-AdvancedFiltersForm__closeButtonContainer closeButtonContainer"
    >
      <WithStyles(ButtonWithHoverFill)
        classes={Object {}}
        displayLabel={false}
        handleClick={[Function]}
        icon="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
        label="Close"
      />
    </Styled(MuiBox)>
    <Styled(MuiBox)
      className="wye-AdvancedFiltersForm__fieldsContainer fieldsContainer"
    >
      <Form>
        <WithStyles(ForwardRef(Typography))
          className="wye-AdvancedFiltersForm__fieldSectionTitle fieldSectionTitle"
          variant="body1"
        >
          Users:
        </WithStyles(ForwardRef(Typography))>
        <Field
          name="users"
        >
          <Component />
        </Field>
        <Styled(MuiBox)
          className="wye-AdvancedFiltersForm__fieldSection fieldSection"
          my={4}
        >
          <WithStyles(ForwardRef(Typography))
            className="wye-AdvancedFiltersForm__fieldSectionTitle fieldSectionTitle"
            variant="body1"
          >
            Game Filters:
          </WithStyles(ForwardRef(Typography))>
          <Styled(MuiBox)
            my={2}
          >
            <Field
              name="filters"
            >
              <Component />
            </Field>
          </Styled(MuiBox)>
          <Styled(MuiBox)
            my={2}
          >
            <Field
              name="filters"
            >
              <Component />
            </Field>
          </Styled(MuiBox)>
          <Styled(MuiBox)
            my={2}
          >
            <Field
              name="filters"
            >
              <Component />
            </Field>
          </Styled(MuiBox)>
          <Styled(MuiBox)
            my={2}
          >
            <Field
              name="filters"
            >
              <Component />
            </Field>
          </Styled(MuiBox)>
          <Styled(MuiBox)
            my={2}
          >
            <Field
              name="filters"
            >
              <Component />
            </Field>
          </Styled(MuiBox)>
          <Styled(MuiBox)
            my={2}
          >
            <Field
              name="filters"
            >
              <Component />
            </Field>
          </Styled(MuiBox)>
          <Styled(MuiBox)
            my={2}
          >
            <Field
              name="filters"
            >
              <Component />
            </Field>
          </Styled(MuiBox)>
          <Styled(MuiBox)
            my={2}
          >
            <Field
              name="filters"
            >
              <Component />
            </Field>
          </Styled(MuiBox)>
        </Styled(MuiBox)>
        <Styled(MuiBox)
          className="wye-AdvancedFiltersForm__fieldSection fieldSection"
          my={4}
        >
          <WithStyles(ForwardRef(Typography))
            className="wye-AdvancedFiltersForm__fieldSectionTitle fieldSectionTitle"
            variant="body1"
          >
            Sorting:
          </WithStyles(ForwardRef(Typography))>
          <Styled(MuiBox)
            my={2}
          >
            <Field
              name="orderBy"
            >
              <Component />
            </Field>
          </Styled(MuiBox)>
          <Styled(MuiBox)
            my={2}
          >
            <Field
              name="sortOrder"
            >
              <Component />
            </Field>
          </Styled(MuiBox)>
        </Styled(MuiBox)>
      </Form>
    </Styled(MuiBox)>
    <Styled(MuiBox)
      className="wye-AdvancedFiltersForm__submitButtonContainer submitButtonContainer"
      my={4}
    >
      <WithStyles(ActionButton)
        classes={Object {}}
        label="Filter"
        onClick={[Function]}
      />
    </Styled(MuiBox)>
  </WithStyles(ForwardRef(Container))>
</ForwardRef(Drawer)>
`;

exports[`AdvancedFiltersFormWithFormik should render successfully 1`] = `
<Formik
  initialValues={
    Object {
      "filters": Object {},
      "orderBy": Array [],
      "sortOrder": "DESC",
      "users": Array [],
    }
  }
>
  <Component />
</Formik>
`;
